### DISCUSSION TITLE
Vi / Vim / Neovim support

### DISCUSSION THREAD

**Discussion by elviskahoro** [warpdotdev/Warp#451] [Archived - Launched]
Vim support

**Comment by elviskahoro** [comment]
### Describe the solution you'd like?

```shell
Vim Speed + Performance similar to alacritty.
```


### Is your feature request related to a problem? Please describe.

```shell
My full Vimrc with various plugins installed, performs noticeably slower in Warp compared to Alacritty. I removed the plugins but kept sensible vim settings from the vimrc to test scrolling time with a base vimrc. Without plugins the difference is not as pronounced and not to the detriment of usability, but with plugins it was very noticeable and led to me using alacritty for text editing related tasks.
```

### Additional context

Results

Alacritty - 59 seconds
08:05:20
08:06:19

Warp - 77 seconds total
08:06:23
08:07:40

~1.3 X - length
Video Here: https://elvis.ai/assets/warp_test_01.m4v
Warp - Test
09/23/2021

Vim's Scrolling Speed via Shift + ]
	is considerably slower in warp compared to alacritty.
	Scroll through a large file (pandas-core/generic.py)

Note: Tried to get the display to be as similar as possible (via alacritty.yml) but the warp window displays 49 vim lines and alacritty displays 48 vim lines.
Have a window running TOP under alacritty and warp for the clock.

Alacritty Settings:
```sh
	window
		padding: x = 8
		padding: y = 0

	font:
		size: 14.0
		offset: y = 2
```
Warp Settings:
```
	font size: 14
```

**Comment by CalgaryMichael** [comment]
x-post from [Discord message](https://discord.com/channels/851854972600451112/852251449068486773/895014131239886929):

Small visual bug that I had noticed when using vim (v 8.2) on Warp. You can see the cursor is bouncing around at random when scrolling on a file. Does not seem to have any affect besides visual -- when you stop scrolling, the cursor is in the correct location.

https://user-images.githubusercontent.com/8285462/136083815-77e90a1a-0818-427c-9630-a593c89706b5.mov

Let me know if you need another further information (vim plugins, etc.)

**Comment by S15** [comment]
Ran into similar performance issues using with vim + warp. I see slow scrolling, slow cursor movement, and slow input. I was directed to this ticket to add information. What I am seeing _appears_ to be related to having a (drawing a?) background color.

I `set termguicolors`, and use a custom color scheme. I link (most of) the rest of the highlight groups against the Normal group, and I set the Normal highlight group to:

```viml
hi Normal guifg=#bbbbbb guibg=#333333 gui=NONE
```
Changing Normal to: 
```viml
hi Normal guifg=#bbbbbb guibg=NONE gui=NONE
```

seems to resolve the performance problems. (Similarly I tried editing with `vim -u NONE` and that seemed to work fine)

**Comment by Caprowni** [comment]
Just wanted to chime in on this after experiencing the same issues with lag using Vim, using the fix @S15 mentioned above worked for me also, speed increased to what I would expect!

**Comment by lrvdijk** [comment]
For me one of the last updates has fixed performance issues.

**Reply by lessp** [reply]
Got prompted with an update today which seems to have solved the performance issue for me as well! 🙌

**Comment by gagata** [comment]
oddly enough i most often hit the issues (gettiing a rainbow circle and all) when pressing `backspace` or switching between modes.

**Comment by elviskahoro** [comment]
From discord #vim-users 11.13.2021
> I can't use ctrl-6 to go back to my latest buffer in nvim. This works in all other terminals I've used so far. 
>  map <C-6> <C-^> even adding this to my init.vim doesn't change anything...
>  <C-^> works but <C-6> doesn't

**Comment by lrvdijk** [comment]
Hi, a few other things that are not entirely correct:

1) I use [Fira Code Nerd Font](https://github.com/ryanoasis/nerd-fonts), together with Vim-airline. The line height seems off, however, and the special glyphs are rendered a bit weirdly.
2) The different color of the window border compared to the vim background color bothers me a bit. Maybe I should configure my vim differently?

<img width="1264" alt="Screen Shot 2021-11-20 at 9 04 13 PM" src="https://user-images.githubusercontent.com/24530/142746612-254d1b78-335c-46f1-bc2f-1c90e6c3c891.png">

**Reply by lessp** [reply]
Also seeing some discrepancies in regards to themes (e.g. notice the line numbers as well as the indentation lines):

**Warp**
![image](https://user-images.githubusercontent.com/17602389/151354809-0333d2e8-877c-4ca8-8ba1-f2cc61a9de42.png)

**Kitty**
![image](https://user-images.githubusercontent.com/17602389/151354740-9796313f-e3b0-4795-bd1d-f7e7c76d312f.png)

**Comment by jmaitoza** [comment]
Here is my neovim output from the Warp terminal:
![image](https://user-images.githubusercontent.com/60358495/143326271-dd14b3fd-3751-4639-93df-6388c88ba7da.png)

same exact prompt from iTerm:
![image](https://user-images.githubusercontent.com/60358495/143326322-bdc5c59d-f72c-42e7-a015-b81a7da3f127.png)


I will note that I've been trying out neovim using [this](https://nvchad.github.io) setup which does run with a lot of plugins

**Comment by elviskahoro** [comment]
Neovim:
"Specifically, Warp seems to struggle with mouse support for neovim. Clicking to move the cursor and scrolling using the touchpad are both pretty unreliable and jittery."

**Comment by elviskahoro** [comment]
Likning:
- https://github.com/warpdotdev/warp/issues/342

**Comment by elviskahoro** [comment]
Linking:
- https://github.com/warpdotdev/warp/issues/376

**Comment by elviskahoro** [comment]
Linking:
- https://github.com/warpdotdev/warp/issues/429

**Comment by lessp** [comment]
There seems to be a rendering issue with the cursor (found when using [`fzf`](https://github.com/junegunn/fzf.vim)), notice the blue selected character in the Warp screenshot.

**Warp:**

![image](https://user-images.githubusercontent.com/17602389/145378609-df571da8-e36e-4523-82d2-f11bef94216b.png)

**Kitty:**

![image](https://user-images.githubusercontent.com/17602389/145378729-94f27f14-0a32-4431-a951-9b15ed8d4f95.png)

**Reply by elviskahoro** [reply]
@charlespierce @gagata

**Comment by elviskahoro** [comment]
"When he is using neovim and the number of lines are shorter than viewport, scrolling down jerks the screen back to the first line"

**Comment by elviskahoro** [comment]
This specific theme makes the cursor disappear. Ported from #368 

For some reason I don't see any cursor when using `vim`. While it's a nice throwback to only navigate using the `line:col` output in the status bar it's a bit too much these days.

The issue disappears when I disable the use of my `vim` theme - [vividchalk](https://github.com/tpope/vim-vividchalk).

Screenshot from both iterm2 and warp attached.

Iterm2
<img width="602" alt="368-iterm" src="https://user-images.githubusercontent.com/29553206/145742309-c6dc3bd4-c6d7-4573-af21-c6514a6e8ddb.png">

Warp
<img width="639" alt="368-warp" src="https://user-images.githubusercontent.com/29553206/145742323-365f5bf4-d585-40d3-9373-f6e01c248d02.png">

**Reply by elviskahoro** [reply]
@madisp

**Comment by elviskahoro** [comment]
From Andrew
One thing that’s a bit funny in neovim, is that warp adds tilde’s under the file in lieu of line numbers
![image](https://user-images.githubusercontent.com/29553206/150025680-84b171bf-d2a9-46b4-986c-76a58f0fa164.png)
Compared with alacritty 
![image](https://user-images.githubusercontent.com/29553206/150025691-46f52256-b988-4173-b8ae-b1b49f13dd60.png)
Also it seems that relative line numbers aren’t working in warp. The above screws are the same file, same vimrc  config

**Comment by jamesgecko** [comment]
Some of my keybindings that use option don't seem to work in Neovim under Warp. I use `A-h` (and jkl) to switch between splits, and the keybinding works when my option keys are configured in iTerm 2 and Terminal, but I don't have such an option in Warp.

The neovim default `<C-w>h` keybinding works, but not my `A-h` mapping. In my `~/config/nvim/config.lua`:

```lua
local map = vim.api.nvim_set_keymap
map('n', '<A-j>', '<C-w>j', { noremap = true })
map('n', '<A-k>', '<C-w>k', { noremap = true })
map('n', '<A-h>', '<C-w>h', { noremap = true })
map('n', '<A-l>', '<C-w>l', { noremap = true })
```

iTerm2 preference | Terminal preference
---|---
![iterm](https://user-images.githubusercontent.com/44252/150186201-b4a23185-049a-41d9-aa27-d25f25b6b4a3.png)|![terminal](https://user-images.githubusercontent.com/44252/150186212-6bce6258-b1cd-406b-8f5b-bf8288055f76.png)

**Reply by jamesgecko** [reply]
Oops, I didn't even see this option is in the File menu.

**Comment by elviskahoro** [comment]
CTRL-/ does not comment out anymore in my neovim

From @corroded via #376

### Describe the solution you'd like?

```shell
I would hope that all the keybindings I have in my vimrc would have been ported over. It looks like there are some keys that have not ported over properly. I think it should be able to port as the neovim window is technically not the terminal window anymore?
```


### Is your feature request related to a problem? Please describe.

_No response_

### Additional context

I am using this plugin:  https://github.com/tomtom/tcomment_vim

Let me know if additional info is needed or there's already an existing issue (I tried to search but couldn't find any)

**Reply by elviskahoro** [reply]
> ctrl shortcuts: "ctrl " usually comments things out in neovim but it doesn't work in warp.

**Reply by oriming** [reply]
I had the same problem and it was bothering me

**Comment by matsdahlin** [comment]
I recently got access to Warp, and NeoVim is pretty much unusable in it, it is noticeably laggy entering text compared to my current terminal (Kitty).

Glad that some of you have seen speed improvements, hoping there will be more improvements coming up!

**Comment by elviskahoro** [comment]
From email:
> need padding/margin from top when terminal app (neovim) launched, there is no space atm, looks a bit strange.

**Reply by stevepond** [reply]
I run NVChad, which is neovim and packer and a bunch LSP and plugins.  Seems like none of the plugin paths can be found...

<img width="992" alt="image" src="https://user-images.githubusercontent.com/367629/192071796-ba3d1b72-0b44-4a94-86c9-2a5c90cc6b4e.png">

**Comment by sullimander** [comment]
I'm using neovim with the [navarasu/onedark.nvim](https://github.com/navarasu/onedark.nvim) theme, and it is rendering very different than it is in iTerm2. Any advice on how to fix the issue?

**iTerm2**
![vim-iterm](https://user-images.githubusercontent.com/39138/160457835-49558590-ef00-4df6-8d87-419ca0c29b84.png)

**Warp**
![vim-warp](https://user-images.githubusercontent.com/39138/160457889-680e5b81-6a15-449b-90bc-c73472e5eb4d.png)

**Reply by npupko** [reply]
Same for me with gruvbox theme

**Reply by gagata** [reply]
I think this may be related to Warp forcing certain contrast on the colors

**Reply by zhaob1n** [reply]
same for me

**Reply by paulstey** [reply]
I'm in the same boat... If anyone has a fix, I'd be very grateful!!

**Reply by twandylue** [reply]
same for me

**Reply by d4be4st** [reply]
It is also missing `undercurl` highlight

You can see your `require` is not underlined in warp.

**Reply by ArtursTimofejevs** [reply]
Same here :(

**Reply by robertwijas** [reply]
same here, the lack of underline is especially annoying :) any workarounds?

**Reply by minaeakhalil** [reply]
Does not appear to be fixed yet! I'm having this issue with various themes.

**Reply by elviskahoro** [reply]
@minaeakhalil What's your current configuration for `Settings > Appearance > Text > Enforce Minimum Contrast`?

If the underline is still missing after changing the setting to `Never` could you share a couple themes it works for and a couple themes it doesn't? Will help us drill down what's breaking it.

**Reply by minaeakhalil** [reply]
@elviskahoro - Thanks for following up and thanks for all your effort in making Warp such an amazing product 😄 

The issue persist even when the `Settings > Appearance > Text > Enforce Minimum Contrast` is set to `Never`.

Here are some additional details, hope it would help troubleshooting this issue:
1. Using currently Neovim 0.9 with [LazyVim pre-configuration](https://www.lazyvim.org/)
2. Tested this issue with different themes:
    - [Catpuccin](https://github.com/catppuccin/nvim) (the one in the attached screenshot),
    - Gruvbox
    - Gruvbox-material
    - TokyoNight (default for LazyVim)

In the screenshot, Warp is Top section, Wezterm is Bottom.

<img width="758" alt="Screenshot 2023-05-19 at 10 59 22 AM" src="https://github.com/warpdotdev/Warp/assets/11508105/3a1613ea-29a6-451f-b6cf-73cfb2e22a32">

**Reply by minaeakhalil** [reply]
@elviskahoro - FYI, this is no longer an issue with the latest stable version of Warp (v0.2023.06.27.19.34.stable_01).

Thanks for the help and the support with Warp 😄

**Comment by nfwyst** [comment]
ligatures font support please

**Comment by colepeters** [comment]
Definitely some massive differences in colour rendering for me. Here's my Neovim in iTerm:
<img width="1130" alt="image" src="https://user-images.githubusercontent.com/1713932/194602250-e710e597-6b36-4cf8-8331-bde1051e0d9d.png">

…and in Warp:
<img width="1136" alt="image" src="https://user-images.githubusercontent.com/1713932/194602564-dcf539b9-2ba5-4e63-92c8-e5e9596afcfd.png">

**Reply by bfulop** [reply]
This should be fixed in the coming weeks:
https://github.com/warpdotdev/Warp/issues/1110#issuecomment-1265638815

**Reply by nyngwang** [reply]
off-topic: What's the name of your colorscheme?

**Comment by ddbrierton** [comment]
This thread is very old now, I have no idea if anyone from Warp is still monitoring it. Anyway, first things first, I’ve been using Warp every day since November 2022 and I absolutely love it. It’s easily the most innovative terminal application I’ve ever used. But I now find myself in the absurd situation of having to have Alacritty open alongside Warp solely so that I can use Neovim.

Whilst many of the earlier issues seem to have been addressed, two of the most fundamental things haven’t:

1. Warp does not relinquish styling of the cursor to Neovim, with the result that in Neovim’s normal mode the character the cursor is on is unreadable.
2. The `option` key on my Mac keyboard does not function properly — this is **with** Settings > Features > Keys > Left/Right Option key is Meta enabled. _Some_ keys seem to work, but keys which in macOS normally function as dead keys, such as `option + i` do not. This seems to be a regression, which I reported in #3018. I use a British keyboard and the `#` character is `option + 3` which does nothing in Neovim running in Warp (but does work in the Warp command line). I write a lot of CSS - the `#` character is pretty essential.

Furthermore, some character decorations such as underlining and wavy underlining are not working in Warp but are in Alacritty, although I don’t consider that an essential feature, but it would be nice to have.

Screenshot of Warp:

<img width="393" alt="Screenshot 2023-05-18 at 13 45 48" src="https://github.com/warpdotdev/Warp/assets/7255543/b87d2d1b-1e92-4e86-91c3-44a03effa006">

Screenshot of Alacritty:

<img width="414" alt="Screenshot 2023-05-18 at 13 46 14" src="https://github.com/warpdotdev/Warp/assets/7255543/7b68439d-7de1-404d-8b01-b6cbd3195ea0">

On the plus side, in my subjective opinion I think that fonts and colours are rendered better in Warp than in Alacritty.

I have to say though that I find it truly disheartening that you appear to be using time and resources adding things like AI when such fundamental things as a usable cursor and keyboard appear to be overlooked. Warp sure is clever, but if I want to get some work done I have to use Alacritty.

Finally, is there a vision for how Warp should handle apps which takeover the full terminal screen? After all, once something like nvim is running everything special about Warp is no longer available. It strike me that there is room to innovate here just as you did with blocks. Perhaps full terminal screen apps should detach into a s separate frame? I hope there is a vision, but please please fix the cursor and the keyboard before working on that!

**Comment by karambaq** [comment]
With line-height more that `1.3` symbols looks like this:

<img width="1167" alt="image" src="https://github.com/warpdotdev/Warp/assets/14233969/5b9de599-daec-4a33-9272-2fccd4e92256">

While in kitty or iTerm2 it looks fine:

<img width="1116" alt="image" src="https://github.com/warpdotdev/Warp/assets/14233969/30cb792d-cc16-44d0-9396-5296731e44c1">

**Comment by Manubi** [comment]
Are there also any plans to just add vim for the normal warp terminal? not the neovim support?
Because I am used to use vim navigation with iTerm. Would be nice to navigate with vim as well.

**Comment by ratbekn** [comment]
Hi! Font ligatures works in command line, but don't work in Neovim. Am I doing something wrong?
<img width="1840" alt="Screenshot 2023-11-27 at 01 30 39" src="https://github.com/warpdotdev/Warp/assets/61360459/815b4ae0-e45d-4240-a912-959c3c0b774c">

**Comment by chadoh** [comment]
Vim within an ssh session is not supported properly. The command header is still shown, which covers the first three lines of the file or so. Here you can see that the "[do, re, mi](https://www.esl-lounge.com/songs/song-do-re-mi.php)" lines are covered by the command header.

<img width="371" alt="image" src="https://github.com/warpdotdev/Warp/assets/221614/59568a43-0536-4974-aa27-699aa972e97d">

**Comment by NeuSandeZ** [comment]
Is it only for me that the NeoVim doesn't take the full width and hight of warp?

![Screenshot 2024-10-04 at 12 24 05](https://github.com/user-attachments/assets/bbdf948f-f506-4749-878a-929c67df4797)

**Reply by ddbrierton** [reply]
No, but I think that it is normal. Warp's window can be any number of pixels tall and wide, but Neovim I think sets its size in characters not pixels.

**Reply by elviskahoro** [reply]
@NeuSandeZ it has to do with Warp's grid. The Warp team explored expanding it way back but it led to some messiness. The most elegant workaround is to have Warp match the color of your neovim :(

**Reply by 2nthony** [reply]
> @NeuSandeZ it has to do with Warp's grid. The Warp team explored expanding it way back but it led to some messiness. The most elegant workaround is to have Warp match the color of your neovim :(

There are couple issues about the gap(padding) I've seen, so it should be add option to tweak the gap(padding). iTerm has supported, wezterm also supported(window_padding)

IMO: terminal should flex to user, but not user forced to flex terminal, because it is a Terminal

this is day theme for warp, but neovim is using night theme.

<img width="952" alt="image" src="https://github.com/user-attachments/assets/99e8bd40-6685-4d13-b4d3-64a0293b8d41">

**Comment by DominickVale** [comment]
Hey, i used warp for a while, and i only recently tried another ghostty based terminal and was kind of surprised to see how much fluid vim looked. Is there something you think i could try to improve the fluidity of neovim on warp? Going back to warp it really feels like i have "less FPS". Test while scrolling feels much more choppy and not fluid. Also inputs seem slightly slower.
